File: app/src/test/java/com/caracode/whatclothes/main/MainPresenterTest.java
package com.caracode.whatclothes.main;

import android.support.v4.util.Pair;

import com.caracode.whatclothes.R;
import com.caracode.whatclothes.api.response.FiveDayResponse;
import com.caracode.whatclothes.api.response.PhotosResponse;
import com.caracode.whatclothes.common.Constants;
import com.caracode.whatclothes.service.PhotoService;
import com.caracode.whatclothes.service.WeatherService;

import org.joda.time.DateTime;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.Collections;

import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.disposables.CompositeDisposable;

import static junit.framework.Assert.assertTrue;
import static org.mockito.Mockito.reset;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class MainPresenterTest {

    @Mock
    MainView viewMock;
    @Mock
    WeatherService weatherServiceMock;
    @Mock
    PhotoService photoServiceMock;
    @Mock
    private FiveDayResponse fiveDayResponseMock;
    @Mock
    private FiveDayResponse.ThreeHourlyUpdate threeHourlyUpdateMock;
    @Mock
    private PhotosResponse photosResponseMock;
    @Mock
    private com.caracode.whatclothes.api.response.FiveDayResponse.ThreeHourlyUpdate.Main updateMainMock;

    CompositeDisposable networkDisposable = new CompositeDisposable();
    CompositeDisposable viewDisposable = new CompositeDisposable();

    private MainPresenter presenter;

    @Before
    public void before() {
        presenter = new MainPresenter(weatherServiceMock, photoServiceMock, networkDisposable, viewDisposable);
        presenter.create();
        when(viewMock.onFabClick()).thenReturn(Observable.never());
        when(weatherServiceMock.getWeather()).thenReturn(Single.never());
        when(photoServiceMock.getPhotos()).thenReturn(Single.never());
    }

    @Test
    public void testFabClick() {
        when(viewMock.onFabClick()).thenReturn(Observable.just(new Object()));

        presenter.attachView(viewMock);

        verify(viewMock).displayUnimplementedMessage();
    }

    @Test
    public void testWeatherWithDefaultPhotos() {
        final DateTime dateTime = new DateTime(2017, 1, 15, 0, 0);
        final String readableDate = "Sun 15 Jan";
        final double maxTemp = 23.47;
        final double minTemp = 7.21;
        arrangeWeatherMocks(dateTime, minTemp, maxTemp);
        arrangePhotoMocks();

        presenter.attachView(viewMock);

        verify(viewMock).updateUi(MainViewModel.create(Collections.singletonList(
                MainViewModel.DayModel.create(readableDate, minTemp, maxTemp, Constants.BACKUP_PHOTOS.get(0),
                        new Pair<>(R.drawable.jumper, R.drawable.trousers)))));
    }

    @Test
    public void testRestoringState() {
        final DateTime dateTime = new DateTime(2020, 3, 8, 0, 0);
        final String readableDate = "Sun 8 Mar";
        final double maxTemp = 15.19;
        final double minTemp = -3.42;
        arrangeWeatherMocks(dateTime, minTemp, maxTemp);
        arrangePhotoMocks();

        presenter.attachView(viewMock);

        presenter.detachView();
        reset(viewMock);

        presenter.attachView(viewMock);
        verify(viewMock).updateUi(MainViewModel.create(Collections.singletonList(
                MainViewModel.DayModel.create(readableDate, minTemp, maxTemp, Constants.BACKUP_PHOTOS.get(0),
                        new Pair<>(R.drawable.jacket, R.drawable.trousers)))));
    }

    @After
    public void after() {
        presenter.detachView();
        presenter.destroy();
        assertTrue(presenter.isDestroyed());
    }

    private void arrangeWeatherMocks(DateTime dateTime, double minTemp, double maxTemp) {
        when(weatherServiceMock.getWeather()).thenReturn(Single.just(fiveDayResponseMock));
        when(fiveDayResponseMock.threeHourlyUpdates()).thenReturn(Collections.singletonList(threeHourlyUpdateMock));
        when(threeHourlyUpdateMock.dateTime()).thenReturn(dateTime);
        when(threeHourlyUpdateMock.main()).thenReturn(updateMainMock);
        when(updateMainMock.maxTemp()).thenReturn(maxTemp);
        when(updateMainMock.minTemp()).thenReturn(minTemp);
    }

    private void arrangePhotoMocks() {
        when(photoServiceMock.getPhotos()).thenReturn(Single.just(photosResponseMock));
    }
}

File: app/src/test/java/com/caracode/whatclothes/main/MainPresenterTest.kt
package com.caracode.whatclothes.main

import com.caracode.whatclothes.R
import com.caracode.whatclothes.api.response.FiveDayResponse
import com.caracode.whatclothes.api.response.PhotosResponse
import com.caracode.whatclothes.common.Constants
import com.caracode.whatclothes.service.PhotoService
import com.caracode.whatclothes.service.WeatherService

import org.joda.time.DateTime
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.junit.MockitoJUnitRunner

import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.disposables.CompositeDisposable

import junit.framework.Assert.assertTrue
import org.mockito.Mockito.reset
import org.mockito.Mockito.verify
import org.mockito.Mockito.`when`

@RunWith(MockitoJUnitRunner::class)
class MainPresenterTest {

    @Mock
    private lateinit var viewMock: MainView
    @Mock
    private lateinit var weatherServiceMock: WeatherService
    @Mock
    private lateinit var photoServiceMock: PhotoService
    @Mock
    private lateinit var fiveDayResponseMock: FiveDayResponse
    @Mock
    private lateinit var threeHourlyUpdateMock: FiveDayResponse.ThreeHourlyUpdate
    @Mock
    private lateinit var photosResponseMock: PhotosResponse
    @Mock
    private lateinit var updateMainMock: com.caracode.whatclothes.api.response.FiveDayResponse.ThreeHourlyUpdate.Main

    private var networkDisposable = CompositeDisposable()
    private var viewDisposable = CompositeDisposable()

    private lateinit var presenter: MainPresenter

    @Before
    fun before() {
        presenter = MainPresenter(weatherServiceMock, photoServiceMock, networkDisposable, viewDisposable)
        presenter.create()
        `when`(viewMock.onFabClick()).thenReturn(Observable.never())
        `when`(weatherServiceMock.weather).thenReturn(Single.never())
        `when`(photoServiceMock.photos).thenReturn(Single.never())
    }

    @Test
    fun testFabClick() {
        `when`(viewMock.onFabClick()).thenReturn(Observable.just(Any()))

        presenter.attachView(viewMock)

        verify<MainView>(viewMock).displayUnimplementedMessage()
    }

    @Test
    fun testWeatherWithDefaultPhotos() {
        val dateTime = DateTime(2017, 1, 15, 0, 0)
        val readableDate = "Sun 15 Jan"
        val maxTemp = 23.47
        val minTemp = 7.21
        arrangeWeatherMocks(dateTime, minTemp, maxTemp)
        arrangePhotoMocks()

        presenter.attachView(viewMock)

        verify<MainView>(viewMock).updateUi(MainViewModel(listOf(DayModel(readableDate, minTemp, maxTemp, Constants.BACKUP_PHOTOS[0],
                Pair(R.drawable.jumper, R.drawable.trousers)))))
    }

    @Test
    fun testRestoringState() {
        val dateTime = DateTime(2020, 3, 8, 0, 0)
        val readableDate = "Sun 8 Mar"
        val maxTemp = 15.19
        val minTemp = -3.42
        arrangeWeatherMocks(dateTime, minTemp, maxTemp)
        arrangePhotoMocks()

        presenter.attachView(viewMock)

        presenter.detachView()
        reset<MainView>(viewMock)

        presenter.attachView(viewMock)
        verify<MainView>(viewMock).updateUi(MainViewModel(listOf(DayModel(readableDate, minTemp, maxTemp, Constants.BACKUP_PHOTOS[0],
                Pair(R.drawable.coat, R.drawable.trousers)))))
    }

    @After
    fun after() {
        presenter.detachView()
        presenter.destroy()
        assertTrue(presenter.isDestroyed)
    }

    private fun arrangeWeatherMocks(dateTime: DateTime, minTemp: Double, maxTemp: Double) {
        `when`(weatherServiceMock.weather).thenReturn(Single.just(fiveDayResponseMock))
        `when`(fiveDayResponseMock.threeHourlyUpdates()).thenReturn(listOf(threeHourlyUpdateMock))
        `when`(threeHourlyUpdateMock.dateTime()).thenReturn(dateTime)
        `when`(threeHourlyUpdateMock.main()).thenReturn(updateMainMock)
        `when`(updateMainMock.maxTemp()).thenReturn(maxTemp)
        `when`(updateMainMock.minTemp()).thenReturn(minTemp)
    }

    private fun arrangePhotoMocks() {
        `when`(photoServiceMock.photos).thenReturn(Single.just(photosResponseMock))
    }
}
